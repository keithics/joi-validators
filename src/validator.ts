import { snakeToCapitalize } from '@keithics/core/lib/helpers';

interface ValidateSchemaInterface {
  error: string;
  value?: unknown;
}

export function validateSchemaId(data, val): ValidateSchemaInterface {
  const { error, value } = data.schema.validate(val, {
    stripUnknown: true,
    abortEarly: false,
    label: false,
    errors: {
      wrap: { label: false, array: '', string: '' },
      segment: false,
    },
  });
  const valid = error == null;

  if (valid) {
    return { error: null, value };
  } else {
    const { details } = error;
    const { message, path } = details[0];

    // get custom error if exists else use joi errors
    // eslint-disable-next-line no-prototype-builtins
    return {
      error: data.errors?.hasOwnProperty(path[0])
        ? data.errors[path[0]]
        : snakeToCapitalize(message),
    };
  }
}
/**
 * Validate express body or query
 * This will strip all non defined fields
 * @param data Schema
 * @param query true if /:id GET queries, otherwise req.body
 */

const validator = (data, query = false) => {
  return function (req, res, next) {
    const val = query ? req.params : req.body;
    const { error: errorMessage, value } = validateSchemaId(data, val);

    if (value) {
      req.body = value;
      next(null, value);
    } else {
      res.status(422).json({ message: errorMessage });
    }
  };
};

export default validator;
