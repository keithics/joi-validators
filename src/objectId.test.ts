import { validateSchemaId } from './validator';
import Joi from './objectId';
import { objectIdValidator } from './core.validator';

export const objectIdValidatorRequired = {
  schema: Joi.object({
    sellerId: Joi.objectId().required(),
  }),
};
export const StringSchema = {
  schema: Joi.object({
    firstName: Joi.string().required(),
  }),
};

const ArraySchema = {
  schema: Joi.object({
    list: Joi.array()
      .min(1)
      .items({
        keywords: Joi.string().required(),
        fallbackUrl: Joi.string().uri().required(),
        weight: Joi.number().required(),
      })
      .required(),
  }),
};

describe('Ark Error Handler Tests', () => {
  test('INVALID object ID - should return error if objectID is invalid', () => {
    expect(validateSchemaId(objectIdValidator, { id: '000' })).toMatchSnapshot();
  });

  test('NO object ID - should return error if objectID is invalid', () => {
    expect(validateSchemaId(objectIdValidatorRequired, { notAnId: '000' })).toMatchSnapshot();
  });

  test('INVALID object ID - should return error if sellerId is invalid', () => {
    expect(validateSchemaId(objectIdValidatorRequired, { sellerId: '0' })).toMatchSnapshot();
  });

  test('VALID object ID - should return error if objectID is invalid', () => {
    expect(
      validateSchemaId(objectIdValidatorRequired, { sellerId: '0'.repeat(24) })
    ).toMatchSnapshot();
  });

  test('VALID String - should return error friendly message', () => {
    expect(validateSchemaId(StringSchema, {})).toMatchSnapshot();
  });

  test('INVALID Array with no values- should return error friendly message', () => {
    expect(validateSchemaId(ArraySchema, {})).toMatchSnapshot();
  });

  test('INVALID Array - should return error friendly message', () => {
    expect(
      validateSchemaId(ArraySchema, {
        list: [
          {
            keywords: 'keywords',
            fallbackUrl: 'http://aaa.com',
          },
        ],
      })
    ).toMatchSnapshot();
  });

  test('INVALID Array with missing item value - should return error friendly message', () => {
    expect(
      validateSchemaId(ArraySchema, {
        list: [
          {
            keywords: '',
            fallbackUrl: 'http://aaa.com',
          },
        ],
      })
    ).toMatchSnapshot();
  });

  test('INVALID Array with missing item value at index 1 - should return error friendly message', () => {
    expect(
      validateSchemaId(ArraySchema, {
        list: [
          {
            keywords: 'keyword',
            fallbackUrl: 'http://aaa.com',
            weight: 1,
          },
          {
            keywords: '',
            fallbackUrl: '',
            weight: null,
          },
        ],
      })
    ).toMatchSnapshot();
  });
});
