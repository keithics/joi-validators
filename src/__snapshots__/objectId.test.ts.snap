// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ark Error Handler Tests INVALID Array - should return error friendly message 1`] = `
Object {
  "error": "List weight is required",
}
`;

exports[`Ark Error Handler Tests INVALID Array with missing item value - should return error friendly message 1`] = `
Object {
  "error": "List keywords is not allowed to be empty",
}
`;

exports[`Ark Error Handler Tests INVALID Array with missing item value at index 1 - should return error friendly message 1`] = `
Object {
  "error": "List keywords is not allowed to be empty",
}
`;

exports[`Ark Error Handler Tests INVALID Array with no values- should return error friendly message 1`] = `
Object {
  "error": "List is required",
}
`;

exports[`Ark Error Handler Tests INVALID object ID - should return error if objectID is invalid 1`] = `
Object {
  "error": "Id is an invalid  Object Id with a value of - 000",
}
`;

exports[`Ark Error Handler Tests INVALID object ID - should return error if sellerId is invalid 1`] = `
Object {
  "error": "Seller Id is an invalid  Object Id with a value of - 0",
}
`;

exports[`Ark Error Handler Tests NO object ID - should return error if objectID is invalid 1`] = `
Object {
  "error": "Seller Id is required",
}
`;

exports[`Ark Error Handler Tests VALID String - should return error friendly message 1`] = `
Object {
  "error": "First Name is required",
}
`;

exports[`Ark Error Handler Tests VALID object ID - should return error if objectID is invalid 1`] = `
Object {
  "error": null,
  "value": Object {
    "sellerId": "000000000000000000000000",
  },
}
`;
